// ngx-draggable-resize/esm2015/lib/models/position.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function IPosition() { }
if (false) {
    /** @type {?} */
    IPosition.prototype.x;
    /** @type {?} */
    IPosition.prototype.y;
    /** @type {?|undefined} */
    IPosition.prototype.rotate;
    /** @type {?|undefined} */
    IPosition.prototype.scaleX;
    /** @type {?|undefined} */
    IPosition.prototype.scaleY;
}
export class Position {
    /**
     * @param {?} x
     * @param {?} y
     * @param {?=} rotate
     * @param {?=} scaleX
     * @param {?=} scaleY
     */
    constructor(x, y, rotate = 0, scaleX = 1, scaleY = 1) {
        this.x = x;
        this.y = y;
        this.rotate = rotate;
        this.scaleX = scaleX;
        this.scaleY = scaleY;
    }
    /**
     * @param {?} e
     * @param {?=} el
     * @return {?}
     */
    static fromEvent(e, el = null) {
        /**
         * Fix issue: Resize doesn't work on Windows10 IE11 (and on some windows 7 IE11)
         * https://github.com/xieziyu/ngx-draggable-resize/issues/164
         * e instanceof MouseEvent check returns false on IE11
         */
        if (this.isMouseEvent(e)) {
            return new Position(e.clientX, e.clientY);
        }
        else {
            if (el === null || e.changedTouches.length === 1) {
                return new Position(e.changedTouches[0].clientX, e.changedTouches[0].clientY);
            }
            /**
             * Fix issue: Multiple phone draggables at the same time
             * https://github.com/xieziyu/ngx-draggable-resize/issues/128
             */
            for (let i = 0; i < e.changedTouches.length; i++) {
                if (e.changedTouches[i].target === el) {
                    return new Position(e.changedTouches[i].clientX, e.changedTouches[i].clientY);
                }
            }
        }
    }
    /**
     * @param {?} e
     * @return {?}
     */
    static isMouseEvent(e) {
        return Object.prototype.toString.apply(e).indexOf("MouseEvent") === 8;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    static isIPosition(obj) {
        return !!obj && "x" in obj && "y" in obj;
    }
    /**
     * @param {?} el
     * @return {?}
     */
    static getCurrent(el) {
        /** @type {?} */
        let pos = new Position(0, 0, 0, 0, 0);
        if (window) {
            /** @type {?} */
            const computed = window.getComputedStyle(el);
            if (computed) {
                /** @type {?} */
                let x = parseInt(computed.getPropertyValue("left"), 10);
                /** @type {?} */
                let y = parseInt(computed.getPropertyValue("top"), 10);
                pos.x = isNaN(x) ? 0 : x;
                pos.y = isNaN(y) ? 0 : y;
            }
            return pos;
        }
        else {
            console.error("Not Supported!");
            return null;
        }
    }
    /**
     * @param {?} p
     * @return {?}
     */
    static copy(p) {
        return new Position(0, 0, 0, 1, 1).set(p);
    }
    /**
     * @return {?}
     */
    get value() {
        return {
            x: this.x,
            y: this.y,
            rotate: this.rotate,
            scaleX: this.scaleX,
            scaleY: this.scaleY
        };
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} p
     * @return {THIS}
     */
    add(p) {
        (/** @type {?} */ (this)).x += p.x;
        (/** @type {?} */ (this)).y += p.y;
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} p
     * @return {THIS}
     */
    subtract(p) {
        (/** @type {?} */ (this)).x -= p.x;
        (/** @type {?} */ (this)).y -= p.y;
        return (/** @type {?} */ (this));
    }
    /**
     * @param {?} n
     * @return {?}
     */
    multiply(n) {
        this.x *= n;
        this.y *= n;
    }
    /**
     * @param {?} n
     * @return {?}
     */
    divide(n) {
        this.x /= n;
        this.y /= n;
    }
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    reset() {
        (/** @type {?} */ (this)).x = 0;
        (/** @type {?} */ (this)).y = 0;
        (/** @type {?} */ (this)).rotate = 0;
        (/** @type {?} */ (this)).scaleX = 1;
        (/** @type {?} */ (this)).scaleY = 1;
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} p
     * @return {THIS}
     */
    set(p) {
        (/** @type {?} */ (this)).x = p.x;
        (/** @type {?} */ (this)).y = p.y;
        (/** @type {?} */ (this)).rotate = p.rotate;
        (/** @type {?} */ (this)).scaleY = p.scaleY;
        (/** @type {?} */ (this)).scaleX = p.scaleX;
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    Position.prototype.x;
    /** @type {?} */
    Position.prototype.y;
    /** @type {?} */
    Position.prototype.rotate;
    /** @type {?} */
    Position.prototype.scaleX;
    /** @type {?} */
    Position.prototype.scaleY;
}
